File: mobile/lib/core/geo.dart
import 'dart:math' as math;
import 'package:geolocator/geolocator.dart';

/// Returns the device's current (lat, lng), or null if location services
/// are off or permission was denied - mirrors the web app's
/// `navigator.geolocation.getCurrentPosition` usage (delivery/index.js,
/// ride-sharing/index.js). Plain GPS only - address-to-coordinates
/// geocoding is handled separately by core/places.dart's Places
/// Autocomplete/Details calls.
Future<(double, double)?> getCurrentLatLng() async {
  final serviceEnabled = await Geolocator.isLocationServiceEnabled();
  if (!serviceEnabled) return null;

  var permission = await Geolocator.checkPermission();
  if (permission == LocationPermission.denied) {
    permission = await Geolocator.requestPermission();
  }
  if (permission == LocationPermission.denied ||
      permission == LocationPermission.deniedForever) {
    return null;
  }

  final position = await Geolocator.getCurrentPosition(
    locationSettings: const LocationSettings(accuracy: LocationAccuracy.high),
  );
  return (position.latitude, position.longitude);
}

const _earthRadiusKm = 6371.0;

double _toRad(double deg) => deg * math.pi / 180;

/// Great-circle distance in km between two lat/lng points (Haversine
/// formula) - Dart port of server/src/utils/geo.js's haversineDistanceKm,
/// used by the delivery tracking screen to show "X km from the dropoff".
double haversineDistanceKm(double lat1, double lng1, double lat2, double lng2) {
  final dLat = _toRad(lat2 - lat1);
  final dLng = _toRad(lng2 - lng1);
  final a = math.pow(math.sin(dLat / 2), 2) +
      math.cos(_toRad(lat1)) * math.cos(_toRad(lat2)) * math.pow(math.sin(dLng / 2), 2);
  final c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));
  return _earthRadiusKm * c;
}

File: server/src/utils/geo.js
const EARTH_RADIUS_KM = 6371;

function toRad(deg) {
  return (deg * Math.PI) / 180;
}

/** Great-circle distance in km between two lat/lng points (Haversine formula). */
function haversineDistanceKm(lat1, lng1, lat2, lng2) {
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_KM * c;
}

/**
 * True when all four coordinates are present and usable for a real
 * distance calculation. Both pickup and dropoff need real coordinates -
 * this backend never geocodes an address itself, so a typed address only
 * has coordinates if the client supplied them directly: either via the
 * browser's Geolocation API (rideshare, and delivery's "use my location"),
 * or by picking a real suggestion from delivery's Google Places
 * Autocomplete field (see AddressAutocompleteField.js), which resolves to
 * a geocoded lat/lng client-side before the request ever reaches here.
 */
function hasCoordinates(...values) {
  return values.every((v) => typeof v === "number" && Number.isFinite(v));
}

module.exports = { haversineDistanceKm, hasCoordinates };
